File: benchmarks/executable/restaurant_uint_masked_shift_right.c
// C23 reference for the Restaurant UInt masked-shift-right workload.

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_value = UINT64_C(128);
static volatile uint64_t runtime_count = UINT64_C(65);

static uint64_t masked_shift_right_u64(uint64_t value, uint64_t count) {
  const uint64_t masked_count = count & UINT64_C(63);
  return value >> masked_count;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  const uint64_t masked =
      masked_shift_right_u64(runtime_value, runtime_count);
  return printf("Masked %" PRIu64 "\n", masked) < 0 ? 1 : 0;
}

File: benchmarks/executable/restaurant_uint_masked_shift_right.rs
// Rust 2024 reference for the Restaurant UInt masked-shift-right workload.

use std::hint::black_box;
use std::io::{self, Write};

fn masked_shift_right_u64(value: u64, count: u64) -> u64 {
    value.wrapping_shr(count as u32)
}

fn main() {
    let masked = masked_shift_right_u64(
        black_box(128_u64),
        black_box(65_u64),
    );
    let mut stdout = io::stdout().lock();
    if writeln!(stdout, "Masked {masked}").is_err() {
        std::process::exit(1);
    }
}
